starfish/image/_base.py

```diff
@@ -1,32 +1,77 @@
 import collections
-from typing import Tuple
+from typing import Optional, Sequence, Tuple
+
+from numpy import ndarray
+
+from starfish.constants import Indices
 
 
 class ImageBase(object):
     @classmethod
-    def from_url(cls, relativeurl, baseurl):
+    def from_url(cls, url: str, baseurl: Optional[str]) -> "ImageBase":
+        """
+        Constructs an ImageBase object from a URL and a base URL.
+
+        Parameters:
+        -----------
+        url : str
+            Either an absolute URL or a relative URL referring to the image to be read.
+        baseurl : Optional[str]
+            If url is a relative URL, then this must be provided.  If url is an absolute URL, then this parameter is
+            ignored.
+        """
         raise NotImplementedError()
 
     @property
-    def numpy_array(self):
-        """Retrieves the image data as a numpy array."""
+    def numpy_array(self) -> ndarray:
+        """Retrieves a view of the image data as a numpy array."""
+        raise NotImplementedError()
+
+    @numpy_array.setter
+    def numpy_array(self, data: ndarray):
+        """Sets the image's data from a numpy array.  The numpy array is advised to be immutable afterwards."""
         raise NotImplementedError()
 
     @property
     def raw_shape(self) -> Tuple[int]:
-        """Retrieves the shape of the image data, as a list of the sizes of the indices."""
+        """
+        Returns the shape of the space that this image inhabits.  It does not include the dimensions of the image
+        itself.  For instance, if this is an X-Y image in a C-H-Y-X space, then the shape would include the dimensions C
+        and H.
+
+        Returns
+        -------
+        A list of the sizes of the indices.
```

Comment:
I think `list` could be confusing here. I like the numpydoc return codes. Here it would be: 
```
Returns
--------
Tuple[int] :
    A list of the sizes of the indices
```

It's potentially worth discussing how we want to do type annotations. The information in the docs is made somewhat irrelevant by having them in the function definitions. I suspect that the IDEs probably prefer one over the other, so there is a chance that we'd introduce errors where they don't line up with one another. 

As a dev, I like the function definitions to be typed. As a user, I think I'd probably prefer the docstrings. 🙃